### ISSUE TITLE
[ray] How to write into numpy arrays in shared memory with Ray?

### ISSUE THREAD

**Issue by monocongo** [ray-project/ray#6507]
I am attempting to rewrite Python multiprocessing code using Ray since it appears to be able to abstract shared memory management issues and perform parallel computation faster than straight multiprocessing (based on [this article](https://towardsdatascience.com/10x-faster-parallel-python-without-python-multiprocessing-e5017c93cce1)). My goal is to process all timeseries for a lat/lon grid (with both input and output arrays having shape [lat, lon, time]) in parallel without unnecessary copies of the input/output arrays. The idea is to have both input and output arrays in shared memory and multiple processes will read and write into the shared memory arrays so no copies/serialization are needed for each process to access the arrays being worked upon.

My use case is that I have a CPU-heavy function that I want to apply on all 1-D sub-arrays of a 3-D array. I have managed to make this execute much faster using [a home-rolled approach for shared memory objects with multiprocessing](https://github.com/monocongo/climate_indices/blob/master/climate_indices/__main__.py), but the code is much more convoluted/complicated than I'm comfortable with and I'm hoping to simplify it using ray. However I've not yet worked out how to write into shared memory with ray, and without that, I don't see how this can be done. Hopefully someone reading this can suggest a solution.

I have a [Jupyter notebook](https://nbviewer.jupyter.org/gist/monocongo/35328ef88af9f4ff8a0eb7e5f6a2854b) with a simple example of what I've tried to get this to work using ray.

Here's the gist:

I initialize my environment for ray and create a function that performs a simple operation on a 1-D slice of a 3-D array and writes the result into an output array, with the expectation that this function can run in parallel and read/write on shared memory representations of my input/output arrays:
```python
import psutil
import numpy as np
import ray

num_cpus = psutil.cpu_count(logical=False)
ray.init(num_cpus=num_cpus, ignore_reinit_error=True)

@ray.remote
def add_average_ray(
        in_ary: np.ndarray,
        out_ary: np.ndarray,
        lat_index: int,
        lon_index: int,
):
    ary = in_ary[lat_index, lon_index]
    out_ary[lat_index, lon_index] = ary + np.mean(ary)
```

Next, I create a function that will loop over a 3-D grid of values and apply the above function to each in parallel using ray:
```python
def compute_with_ray(
        input_array: np.ndarray,
) -> np.ndarray:
    # create an output array that computed values will be written
    output_array = np.full(shape=input_array.shape, fill_value=np.NaN)

    # put the input and output arrays into ray's object store
    in_array_id = ray.put(input_array)
    out_array_id = ray.put(output_array)

    # make a list of futures, one per lat/lon (assuming shape (lat, lon, time))
    futures = []
    for lat_index in range(input_array.shape[0]):
        for lon_index in range(input_array.shape[1]):
            futures.append(add_average_ray.remote(in_array_id, out_array_id, lat_index, lon_index))

    # launch the remote tasks in parallel
    ray.get(futures)

    return output_array
```
Next I make a an input array and exercise the code:
```python 
# create an array that can be used to represent a 2x2 cell lat/lon map with 3 times
tst_ary = np.array([[[1, 6, 5], [3, 2, 7]], [[8, 4., 6.], [9, 4, 2]]])

# exercise the ray remote function in parallel
average_added_ray = compute_with_ray(tst_ary)
```
Apparently this is not possible since the arrays that have been added into ray's object store are read-only, and it results in an error:
```
RayTaskError(ValueError): ray_worker (pid=5260, host=skypilot)
  File "<ipython-input-5-0bf9c2bf3f2e>", line 9, in add_average_ray
ValueError: assignment destination is read-only
```

Is there a better way to approach/accomplish this parallel processing on numpy arrays using ray? Thanks in advance for any insight or suggestions.

**Comment by robertnishihara** [comment]
You're right that objects placed in the object store are immutable (to prevent errors arising from multiple processes trying to mutate the same objects).

To have correct behavior, you should create a copy of the array before mutating it.

If you really want to mutate the array in place and if you are only using 1 machine (but keep in mind this can lead to incorrect/unexpected behavior depending on what you're trying to do), you can set `x.flags.writeable = True`, where `x` is the array. This is not a supported part of the Ray API, but technically it should work.

But instead of passing in the output array, I'd suggest just returning the array and then concatenating them, since that will work with all immutable objects.

**Comment by monocongo** [comment]
Thanks for your help, @robertnishihara. 

I have [modified my example code](https://nbviewer.jupyter.org/gist/monocongo/0aa4ca95f63dc5627b80a36793fbb64c) to use the approach you've suggested above. Parallel execution with ray runs significantly slower (~10x) than the same code running in a loop.

This seems like a rudimentary, dead-simple exercise in parallelization where ray should shine (pun intended) -- either I'm still doing something wrong or ray isn't suited for this sort of parallel processing. I suspect the problem to be user error, or I have misunderstood the value proposition offered by ray. Where have I gone wrong?

**Comment by simon-mo** [comment]
Hi @monocongo 

You are applying the compute functions on each `1D` slices of the array. this means there are 400*400 = 160000 tasks being scheduled and run. The overhead of running and scheduling task is greater than the actual time used for compute. This is why ray is slower.

For your use case, I suggest chunking the array to `num_cpu` task and only issue that many tasks. or for simplicity, use apply function on each `2D` slices. This should be significantly faster because only 400 tasks will be instantiated and run across your machine.

**Comment by bioinfornatics** [comment]
Dear, I would like to do the same thing i.e share a writeable buffer.
But I do not understande th @robertnishihara answer by doing : `x.flags.writeable = True`

I tried:
```python
import ray
ray.init()
out_data = ['' for _ in range(0, chunk_size)]
out_data_id = ray.put(out_data)
out_data_id.flags.writeable = True
```
which raise this error
```
AttributeError                            Traceback (most recent call last)
<ipython-input-43-eb9b2a797b1b> in <module>
----> 1 out_data_id.flags.writeable = True

AttributeError: 'ray._raylet.ObjectRef' object has no attribute 'flags'K
```

**Comment by sanjaysrikakulam** [comment]
Hi,

I am attempting the following procedure

https://discuss.ray.io/t/altering-numpy-array-stored-in-plasma-store-in-cython-using-the-start-address-of-the-memory/1885

But not sure about the repercussions. Can you please comment?

**Comment by cszhu** [comment]
Hello! This P3 issue has seen no activity in at least a year. It will be closed in 2 weeks as part of ongoing cleanup efforts.

Please remove the `pending-cleanup` label if you believe this issue should remain open.

Thanks for contributing to Ray!
